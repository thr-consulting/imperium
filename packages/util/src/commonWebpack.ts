import path from 'path';
import nodeExternals from 'webpack-node-externals';
import process from 'process';
import inspectLoader from './inspectLoader';

interface CommonWebpackParams {
	isProduction: boolean;
	isClient: boolean;
	name: string;
	entry: string;
	outputFile: string;
	rules?: any[];
}

export default function commonWebpack({isProduction, isClient, name, entry, outputFile, rules}: CommonWebpackParams) {
	return {
		entry,
		mode: isProduction ? 'production' : 'development',
		context: path.resolve(process.cwd(), 'src'),
		target: isClient ? 'web' : 'node',
		devtool: isProduction ? 'source-map' : 'inline-source-map',
		externals: [
			nodeExternals({modulesDir: 'node_modules'}),
			nodeExternals({modulesDir: path.join('..', '..', 'node_modules')}),
		],
		output: {
			filename: isProduction ? `${path.basename(outputFile, path.extname(outputFile))}.min.js` : outputFile,
			path: path.resolve(process.cwd(), 'dist'),
			library: name,
			libraryTarget: 'commonjs2',
		},
		resolve: {
			extensions: ['.js', '.mjs', '.ts', '.d.ts'].concat(isClient ? ['.tsx'] : []),
		},
		optimization: {
			minimize: isProduction,
			nodeEnv: false,
		},
		module: {
			rules: [
				{
					test: /\.js$/,
					exclude: /node_modules/,
					use: [
						inspectLoader('BABEL'),
						{
							loader: 'babel-loader',
							options: {
								babelrc: false,
								presets: [['@imperium/babel-preset-imperium', {client: isClient}]],
							},
						},
					],
				},
				{
					test: isClient ? /\.tsx?$/ : /\.ts$/,
					exclude: /node_modules/,
					use: [
						inspectLoader('BABEL-TS'),
						{
							loader: 'babel-loader',
							options: {
								babelrc: false,
								presets: [['@imperium/babel-preset-imperium', {client: isClient, typescript: true}]],
							},
						},
					],
				},
			].concat(rules || []),
		},
	};
}
