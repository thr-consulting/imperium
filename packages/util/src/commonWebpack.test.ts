import commonWebpack from './commonWebpack';

jest.mock('process', () => ({
	cwd: () => '/',
}));

describe('commonWebpack', () => {
	it('should work in dev/server mode', () => {
		const out = commonWebpack({
			isProduction: false,
			entry: 'entry.ts',
			isClient: false,
			name: 'commonWebpackTest',
			outputFile: 'out.ts',
		});
		expect(out).toMatchSnapshot();
	});

	it('should work in dev/client mode', () => {
		const out = commonWebpack({
			isProduction: false,
			entry: 'entry.ts',
			isClient: true,
			name: 'commonWebpackTest',
			outputFile: 'out.ts',
		});
		expect(out).toMatchSnapshot();
	});

	it('should work in prod/server mode', () => {
		const out = commonWebpack({
			isProduction: true,
			entry: 'entry.ts',
			isClient: false,
			name: 'commonWebpackTest',
			outputFile: 'out.ts',
		});
		expect(out).toMatchSnapshot();
	});

	it('should work in prod/client mode', () => {
		const out = commonWebpack({
			isProduction: true,
			entry: 'entry.ts',
			isClient: true,
			name: 'commonWebpackTest',
			outputFile: 'out.ts',
		});
		expect(out).toMatchSnapshot();
	});
});
