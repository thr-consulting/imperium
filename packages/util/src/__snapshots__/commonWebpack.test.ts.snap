// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commonWebpack should work in dev/client mode 1`] = `
Object {
  "context": "/src",
  "devtool": "inline-source-map",
  "entry": "entry.ts",
  "externals": Array [
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "inspect-loader",
            "options": Object {
              "callback": [Function],
              "loaderName": "BABEL",
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "@imperium/babel-preset-imperium",
                  Object {
                    "client": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "inspect-loader",
            "options": Object {
              "callback": [Function],
              "loaderName": "BABEL-TS",
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "@imperium/babel-preset-imperium",
                  Object {
                    "client": true,
                    "typescript": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "minimize": false,
    "nodeEnv": false,
  },
  "output": Object {
    "filename": "out.ts",
    "library": "commonWebpackTest",
    "libraryTarget": "commonjs2",
    "path": "/dist",
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".mjs",
      ".ts",
      ".d.ts",
      ".tsx",
    ],
  },
  "target": "web",
}
`;

exports[`commonWebpack should work in dev/server mode 1`] = `
Object {
  "context": "/src",
  "devtool": "inline-source-map",
  "entry": "entry.ts",
  "externals": Array [
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "inspect-loader",
            "options": Object {
              "callback": [Function],
              "loaderName": "BABEL",
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "@imperium/babel-preset-imperium",
                  Object {
                    "client": false,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.ts\\$/,
        "use": Array [
          Object {
            "loader": "inspect-loader",
            "options": Object {
              "callback": [Function],
              "loaderName": "BABEL-TS",
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "@imperium/babel-preset-imperium",
                  Object {
                    "client": false,
                    "typescript": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "minimize": false,
    "nodeEnv": false,
  },
  "output": Object {
    "filename": "out.ts",
    "library": "commonWebpackTest",
    "libraryTarget": "commonjs2",
    "path": "/dist",
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".mjs",
      ".ts",
      ".d.ts",
    ],
  },
  "target": "node",
}
`;

exports[`commonWebpack should work in prod/client mode 1`] = `
Object {
  "context": "/src",
  "devtool": "source-map",
  "entry": "entry.ts",
  "externals": Array [
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "inspect-loader",
            "options": Object {
              "callback": [Function],
              "loaderName": "BABEL",
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "@imperium/babel-preset-imperium",
                  Object {
                    "client": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "inspect-loader",
            "options": Object {
              "callback": [Function],
              "loaderName": "BABEL-TS",
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "@imperium/babel-preset-imperium",
                  Object {
                    "client": true,
                    "typescript": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "minimize": true,
    "nodeEnv": false,
  },
  "output": Object {
    "filename": "out.min.js",
    "library": "commonWebpackTest",
    "libraryTarget": "commonjs2",
    "path": "/dist",
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".mjs",
      ".ts",
      ".d.ts",
      ".tsx",
    ],
  },
  "target": "web",
}
`;

exports[`commonWebpack should work in prod/server mode 1`] = `
Object {
  "context": "/src",
  "devtool": "source-map",
  "entry": "entry.ts",
  "externals": Array [
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "inspect-loader",
            "options": Object {
              "callback": [Function],
              "loaderName": "BABEL",
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "@imperium/babel-preset-imperium",
                  Object {
                    "client": false,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.ts\\$/,
        "use": Array [
          Object {
            "loader": "inspect-loader",
            "options": Object {
              "callback": [Function],
              "loaderName": "BABEL-TS",
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "@imperium/babel-preset-imperium",
                  Object {
                    "client": false,
                    "typescript": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "minimize": true,
    "nodeEnv": false,
  },
  "output": Object {
    "filename": "out.min.js",
    "library": "commonWebpackTest",
    "libraryTarget": "commonjs2",
    "path": "/dist",
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".mjs",
      ".ts",
      ".d.ts",
    ],
  },
  "target": "node",
}
`;
